### PR TITLE
TDL-15168: Use `unique_line_item_id` for invoice updates' lines value instead of `id`

### PR DESCRIPTION
**Pull request by harshpatel4crest**
# Description of change
TDL-15168: Use `unique_line_item_id` for invoice updates' lines value instead of `id`
- Used `unique_id` when available as id field or `id`, as it contains the unique values.

# Manual QA steps
 - 
 
# Risks
 - 
 
# Rollback steps
 - revert this branch

### CODE REVIEW

File: tap_stripe/__init__.py [hunk 1]
```diff
@@ -705,6 +705,27 @@ def sync_sub_stream(sub_stream_name, parent_obj, updates=False):
             obj_ad_dict = sub_stream_obj.to_dict_recursive()
 
             if sub_stream_name == "invoice_line_items":
+                # we will get "unique_id" for default API versions older than "2019-12-03"
```
Comment (savan-chovatiya) [inline tap_stripe/__init__.py:708]: Add a general comment here regarding which field's value moves to another field and all or write one sample example for both records(old and new) with changed field values.

File: tap_stripe/__init__.py [hunk 2]
```diff
@@ -705,6 +705,27 @@ def sync_sub_stream(sub_stream_name, parent_obj, updates=False):
...
+                    # if type is invoiceitem, update 'id' field with 'unique_id'
+                    if obj_ad_dict.get("type") == "invoiceitem":
+                        # get invoice_item id
+                        invoice_item_id = obj_ad_dict.get("id")
+                        obj_ad_dict["id"] = unique_id
+                        # update 'invoice_item' with 'id' if not present
+                        if not obj_ad_dict.get("invoice_item"):
+                            obj_ad_dict["invoice_item"] = invoice_item_id
+                    # if type is subscription, update 'id' field with 'unique_id'
+                    if obj_ad_dict.get("type") == "subscription":
```
Comment (dbshah1212) [inline tap_stripe/__init__.py:721]: Can we use elif here? If both conditions will never get true together.

File: tap_stripe/__init__.py [hunk 3]
```diff
@@ -705,6 +705,27 @@ def sync_sub_stream(sub_stream_name, parent_obj, updates=False):
...
             if sub_stream_name == "invoice_line_items":
+                # we will get "unique_id" for default API versions older than "2019-12-03"
+                if updates and obj_ad_dict.get("unique_id"):
+                    # get unique_id
+                    unique_id = obj_ad_dict.get("unique_id")
+                    # if type is invoiceitem, update 'id' field with 'unique_id'
+                    if obj_ad_dict.get("type") == "invoiceitem":
+                        # get invoice_item id
+                        invoice_item_id = obj_ad_dict.get("id")
+                        obj_ad_dict["id"] = unique_id
```
Comment (dbshah1212) [inline tap_stripe/__init__.py:716]: Take this line out of if.

File: tests/test_all_fields.py [hunk 4]
```diff
@@ -548,8 +572,22 @@ def all_fields_test(self, streams_to_test):
                                         f"AssertionError({failure_1})")
 
                                     nested_key = KNOWN_NESTED_MISSING_FIELDS.get(stream, {})
-                                    if self.find_nested_key(nested_key, expected_field_value, field):
-                                        continue
+                                    # Check whether expected_field_value is list or not.
+                                    # If expected_field_value is list then loop through each item of list
+                                    if type(expected_field_value) == list:
```
Comment (dbshah1212) [inline tests/test_all_fields.py:577]: Move this to the separate function with comments?

Comment (harshpatel4crest) [reply tests/test_all_fields.py:590]: Added this code to a separate function.

File: tap_stripe/__init__.py (hunk 2 above)
Comment (harshpatel4crest) [reply tap_stripe/__init__.py]: Updated.

File: tap_stripe/__init__.py (hunk 1 above)
Comment (harshpatel4crest) [reply tap_stripe/__init__.py:708]: Added comment.

**Review by savan-chovatiya — APPROVED** [review]

**Review by dbshah1212 — APPROVED** [review]

File: tests/test_all_fields.py [hunk 5]
```diff
@@ -548,8 +585,14 @@ def all_fields_test(self, streams_to_test):
...
+                                        if self.handle_list_data(expected_field_value, field, nested_key):
+                                            continue
+                                    else:
+                                        if self.find_nested_key(nested_key, expected_field_value, field):
+                                            continue
```
Comment (kspeer825) [inline tests/test_all_fields.py:595]: See comment in PR 133 
https://github.com/singer-io/tap-stripe/pull/133#discussion_r875965092

**Review by kspeer825 — CHANGES_REQUESTED** [review]

File: tests/test_all_fields.py (hunk 5 above)
Comment (harshpatel4crest) [reply tests/test_all_fields.py:595]: @kspeer There were 2 types of nested fields replicated from the API, 1. dict and 2. list. So, here we have handled dict keys in `find_nested_key` and ignore the failing nested fields and for list fields, we loop over the fields and ignore failing nested fields. Previously we had skipped the first level field directly so we added the above change to ignore only failed nested fields.

**Review by KrisPersonal — APPROVED** [review]

**Review by RushT007 — CHANGES_REQUESTED** [review]
CCi jobs failing, please fix the issues.

**Comment by harshpatel4crest** [comment]
> CCi jobs failing, please fix the issues.

Fixed the issue, the build is passing now.

**Review by RushT007 — APPROVED** [review]

**Review by kspeer825 — APPROVED** [review]

**Pull request merged**